### PR TITLE
test reclaim space

### PR DESCRIPTION
**Pull request by follitude**

### CODE REVIEW

File: stability-tester/suite/reclaim_space.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+// NewReclaimSpaceCase creates a ReclaimSpaceCase.
+func NewReclaimSpaceCase(cfg *config.Config) Case {
+	return &ReclaimSpaceCase{
+		cfg: &cfg.Suite.ReclaimSpace,
+	}
+}
+
+func (c *ReclaimSpaceCase) Initialize(ctx context.Context, db *sql.DB) error {
+	c.createTable(db)
+	return nil
```
Comment (choleraehyq) [inline stability-tester/suite/reclaim_space.go]: Do you mean `return c.createTable(db)`?

File: stability-tester/suite/reclaim_space.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+		}
+		log.Infof("[table], loop: %d", i)
+		c.insertTable(db)
+		c.countNum(db)
+		c.dropTable(db)
+		time.Sleep(time.Duration(c.cfg.SleepTime) * time.Second)
+	}
+
+	return nil
+
```
Comment (choleraehyq) [inline stability-tester/suite/reclaim_space.go]: This function will always return nil.

File: stability-tester/suite/reclaim_space.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+	c.createTable(db)
+	return nil
+}
+
+func (c *ReclaimSpaceCase) String() string {
+	return "reclaim_space"
+}
+
+func (c *ReclaimSpaceCase) Execute(db *sql.DB, index int) error {
+	// ReclaimSpaceCase does not support multithreading.
```
Comment (choleraehyq) [inline stability-tester/suite/reclaim_space.go]: I think `is not concurrent safe` is more clear.

File: stability-tester/suite/reclaim_space.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        gold bigint(30) DEFAULT NULL,
+        diamond bigint(30) DEFAULT NULL,
+        binding_diamond bigint(30) DEFAULT NULL,
+        guild_id bigint(30) DEFAULT NULL,
+        timestamp int(11) DEFAULT NULL,
+        summary varchar(128) DEFAULT NULL,
+        reserved_28 varchar(32) DEFAULT NULL,
+        reserved_29 varchar(32) DEFAULT NULL,
+		reserved_30 varchar(32) DEFAULT NULL,
+		reserved_31 varchar(32) DEFAULT NULL,
```
Comment (choleraehyq) [inline stability-tester/suite/reclaim_space.go]: format?

File: stability-tester/suite/reclaim_space.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (c *ReclaimSpaceCase) insertTable(db *sql.DB) {
+	var wg sync.WaitGroup
+	maxProcs := runtime.NumCPU()
+	runtime.GOMAXPROCS(maxProcs)
```
Comment (choleraehyq) [inline stability-tester/suite/reclaim_space.go]: Go runtime will set this by default on Go 1.5+.

File: stability-tester/suite/reclaim_space.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+	for i := 0; i < c.cfg.Concurrency; i++ {
+		wg.Add(1)
+		go func() {
+			for j := 0; j < c.cfg.Items; j++ {
+				c.insert(db)
+			}
+			defer wg.Done()
```
Comment (choleraehyq) [inline stability-tester/suite/reclaim_space.go]: What does this defer mean?

File: stability-tester/suite/reclaim_space.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+		if _, err := db.Exec(sql); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (c *ReclaimSpaceCase) insertTable(db *sql.DB) error {
+	var wg sync.WaitGroup
+	var e error
```
Comment (siddontang) [inline stability-tester/suite/reclaim_space.go]: wg.Add(c.cfg.Concurrency)

File: stability-tester/suite/reclaim_space.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+	}
+
+	sql := fmt.Sprintf(`INSERT INTO %s VALUES %s`, table, strings.Join(values, ","))
+	if _, err := db.Exec(sql); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (c *ReclaimSpaceCase) countRowNum(db *sql.DB) error {
```
Comment (siddontang) [inline stability-tester/suite/reclaim_space.go]: what is this function for? I think you can remove it.

File: stability-tester/suite/reclaim_space.go [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+func (c *ReclaimSpaceCase) String() string {
+	return "reclaim_space"
+}
+
+func (c *ReclaimSpaceCase) Execute(db *sql.DB, index int) error {
+	if index != 0 {
+		return nil
+	}
+
+	for i := 0; i < c.cfg.LoopCount; i++ {
```
Comment (siddontang) [inline stability-tester/suite/reclaim_space.go]: outer will call the Execute circularly, we don't need LoopCount here. 

Btw, the test will block the current goroutine for a very long time and will block following other cases run in this goroutine, I think you should use another separated goroutine to run it and let this function return nil directly.

Comment (follitude) [reply stability-tester/suite/reclaim_space.go]: ok

**Comment by disksing** [comment]
PTAL @siddontang @choleraehyq @cwen0

**Review by zhangjinpeng87 — COMMENTED** [review]
LGTM

File: stability-tester/serial_suite/reclaim_space.go [hunk 10]
```diff
@@ -0,0 +1,369 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (choleraehyq) [inline stability-tester/serial_suite/reclaim_space.go]: s/2016/2017

**Review by choleraehyq — COMMENTED** [review]
Rest LGTM

**Comment by disksing** [comment]
inactive for too long, closing. Reopen if need.

**Pull request closed without merging**